tests/0102-static_group_rebalance.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * librdkafka - Apache Kafka C library
+ *
+ * Copyright (c) 2012-2019, Magnus Edenhill
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "test.h"
+
+/**
+ * @name KafkaConsumer static membership tests
+ *
+ * Runs two consumers subscribing to a topic simulating various
+ * rebalance scenerios.
+ */
+
+typedef struct _consumer_s {
+        rd_kafka_t *rk;
+        int assign_cnt;
+        int max_rebalance_cnt;
+} _consumer_t;
+
+/**
+ * Calls poll to serve any pending callbacks. Blocks up to timeout_s + 100 ms
+ */
+static void do_consume (_consumer_t *cons, int timeout_s) {
+        rd_kafka_message_t *rkm;
+
+        rkm = rd_kafka_consumer_poll(cons->rk, 100+(timeout_s*1000));
+        if (!rkm)
```

Comment:
I think you want to be more stringent here, if you're expecting a message and don't get one, then fail the test.
Tests should not be forgiving.

Reply:
But the test should also verify that messages are received as expected, not only that rebalances happen.
Otherwise we don't have any test coverage for a static member consumer that consumes messages, which is the functionality that we're offering.